pkg/apis/networking/types.go

```diff
@@ -246,16 +246,39 @@ type IngressList struct {
 
 // IngressSpec describes the Ingress the user wishes to exist.
 type IngressSpec struct {
-	// IngressClassName is the name of the IngressClass cluster resource. The
-	// associated IngressClass defines which controller will implement the
-	// resource. This replaces the deprecated `kubernetes.io/ingress.class`
-	// annotation. For backwards compatibility, when that annotation is set, it
-	// must be given precedence over this field. The controller may emit a
-	// warning if the field and annotation have different values.
-	// Implementations of this API should ignore Ingresses without a class
-	// specified. An IngressClass resource may be marked as default, which can
-	// be used to set a default value for this field. For more information,
-	// refer to the IngressClass documentation.
+	// IngressClassName is the name of an IngressClass cluster resource.
+	// Ingress controller implementations use this field to know whether they
+	// should be picking up this Ingress.
+	//
+	// The field `ingressClassName` (referred to as “the field”) should be
+	// preferred over the annotation `kubernetes.io/ingress.class` (referred to
```

Comment:
I think this wording is going to be _very_ confusing. I know you're trying to say that users should prefer to use IngressClassName instead of the annotation, but we also need to be clear that when both are present, controllers should honor the annotation for backwards compatibility.

Reply:
I suppose you are referring to the wording "preferred" instead of "deprecated". My intent was to make it clear that the annotation support is not going away with Ingress v1, and that implementors should not scare people off with "it will be removed" when in reality the annotation support should stay as long as Ingress v1 exists.

Quoting [this message](https://kubernetes.slack.com/archives/C09QYUH5W/p1641818790038500) from @munnerz in the #sig-network channel:

> In ingress-nginx, **it appears there’s an expectation that support for the annotation will be removed altogether when v1beta1 is no longer supported**, but I don’t think this is correct as it’d mean that a v1 resource that a user creates that specifies this annotation will stop working one day. It would seem to be that the only time we can safely remove this behaviour from ingress controllers, is in a v2 API.

I will re-add the term "deprecated". How do you think we should address this?

Reply:
To clarify, I think saying something like "users should use this field by default instead of the annotation" might be a bit clearer. I'm hesitant to include any text that says the field is preferred over the annotation, because that's not how implementations will actually handle this.

>How do you think we should address the question of "when" the annotation will be removed?

I don't think there's ever going to be a way to remove the annotation or prevent it from being set. As I understand it, the annotation appeared somewhat organically and was never properly defined in upstream Kubernetes (until we introduced IngressClass and formally acknowledged the already existing pattern). If I were building a new implementation of the Ingress API tomorrow, I'd completely ignore the annotation. If I had an implementation that already supported the annotation, I might consider a long term roadmap towards deprecating and removing support for it, but I'm not sure what that timeline would look like. If it is not difficult to support, it may be simplest and most user friendly to support the implementation indefinitely.

Reply:
> If I were building a new implementation of the Ingress API tomorrow, I'd completely ignore the annotation.

The issue with this is that any clients (eg cert-manager) that want to make use of the Ingress API in an automated fashion need to be able to use one mechanism for configuring the ingress class that works with all ingress controller implementations.

If a new ingress controller is created and it doesn't utilise the annotation at all, existing clients of the ingress API won't get the behaviour they expected.

The alternative is that client's like cert-manager expose an option to users to set either the annotation OR the ingressClassName - but this is pushing newfound complexity onto end users.

Reply:
@munnerz that's a good point. This transition is unfortunately painful for implementations. It seems like our options here would be:

1. Implementations like cert-manager support populating either field or annotation depending on config
2. All Ingress controllers indefinitely support the annotation

Unfortunately 2 would only be a partial solution, as it would make it impossible for users to use IngressClass resources with cert-manager. We discussed some other options in Slack as well, it may be worth following up with this on the sig-net mailing list to explore any potential alternatives. Sorry the transition from v1beta1 -> v1 has become so complicated to support, hopefully we can avoid these kinds of issues with any future Gateway API updates.

Reply:
Hey, about ingress-nginx I brought this discussion (async, right now) and I'm really tempted to "undeprecate" the annotation as we discussed previously in the thread, so I'm a huge +1 on using the word "recommended" instead of "deprecated".

It happens that, as @munnerz said a lot of ingress implementations still relies on annotations and not on the new field, and deprecating an annotation is not something that "happens".

We've also pointed some cases were this "deprecation" turned migrations harder, so using the proper wording as you did (recommendation instead of deprecation), at least from me is a gain and also help ingress implementations AND controllers that rely on ingress behavior (like cert-manager) to give users an option and a "path of migration".

Reply:
Let's document it as a deprecated annotation (reflecting the current status - see https://kubernetes.io/docs/reference/labels-annotations-taints/#kubernetes-io-ingress-class-deprecated). If we undeprecate it, that'd be the trigger to update the API docs.